### PR TITLE
Merge thruster features into develop branch

### PR DESCRIPTION
**Pull request by kodsurf**
Will initialize pull request. 

Code compiled and runnable. Thruster requests features work the same as in my old branch (feature/thruster_controll).

Code still dirty : further commit would be to improve style and clean the code. 

Please review/advice and lets discuss details ( needs for changes) inside pull request. Fixes would follow with commit reference.

### CODE REVIEW

File: src/mod/l4_thruster.c [hunk 1]
```diff
@@ -69,25 +59,6 @@ const uint16_t CONVERSION[108] = {0,1,2,3,4,5,6,7,8,9,10,
 
 
 
-
-
-const uint8_t REGISTER_VALUES[108] = {0,1,2,3,4,5,6,7,8,9,10,
```
Comment (kodsurf) [inline src/mod/l4_thruster.c]: This is redundant

File: src/mod/l4_thruster.c [hunk 2]
```diff
@@ -190,295 +151,23 @@ void l4_thruster_main (void) {
...
 }
 
 
 
-/////  BELOW WOULD BE FUNCTIONS TO CONTROLL THRUSTER
-
-
-
-
-void SetReservoirTemperature(int argc, char *argv[]){
```
Comment (kodsurf) [inline src/mod/l4_thruster.c]: All manual request functions are deleted. They all can be implemented with general read/set request functions

Except for request to all registers at once

File: src/mod/l4_thruster.c [hunk 3]
```diff
@@ -543,21 +216,21 @@ void ReadAllRegisters(int argc, char *argv[]){
...
-// received buffer of arbitrary size
+
+// received_buffer of arbitrary size
 // len length of actual thruster reply
 void ParseReadRequest(uint8_t* received_buffer,int len){
 
-	uint8_t sender = received_buffer[0];
-	uint8_t receiver = received_buffer[1];
-	uint8_t message_type = received_buffer[2];
+	//uint8_t sender = received_buffer[0];
```
Comment (kodsurf) [inline src/mod/l4_thruster.c:226]: Those variables are unused and prompt warning

File: src/mod/l4_thruster.c [hunk 4]
```diff
@@ -792,114 +461,79 @@ void GeneralSetRequest(int argc, char *argv[]){
...
 		request[2] = MSGTYPE[2]; // READ -3
-		request[3] = 0x00; //checksumm
+		request[3] = 0x00; // checksum
 		request[4] = 0x02; // LENGTH of payload REGISTER and length
-		request[5] = 0x00; //hardcoded because length of payload is defined with two bytes of uint16
-		request[6]= REGISTER_VALUES[access_register]; // Access register at address requested by used
-
-        //printf("\n register value is set to hex %02X  int %d\n",request[6],request[6]);
+		request[5] = 0x00; // Hardcoded because length of payload is defined with two bytes of uint16
+		request[6]= access_register; // Access register at address requested by used
```
Comment (kodsurf) [inline src/mod/l4_thruster.c:524]: no need to use REGISTER_VALUES array. 

long time ago when I initialized it - I thought that requster addresses are not sequential

File: src/mod/l4_thruster.h [hunk 5]
```diff
@@ -12,20 +12,18 @@
...
+// Variables below will store value returned by read single register request
 // Output from READ SINGLE REGISTER can be a variable of either 1,2 or 4 bytes long.
 // Those global variables will represent LATEST RECEIVED VALUE returned by READ SINGLE REGISTER request
 extern uint8_t VALUE_UINT8;
 extern uint16_t VALUE_UINT16;
 extern uint32_t VALUE_UINT32;
 extern uint8_t READ_REQUEST_OK_FLAG;  // if checksum of parsed message was verified, then message is considered valid. Corresponding flag is set to 1
 extern uint8_t LATEST_ACCESSED_REGISTER; // Latest register to which SET or READ request have been made
-extern double ACTUAL_VALUE; // After extracting either uint8 or uint16 or uint32 value to obtain actual real value conversion multipliers have to be applied
+extern double ACTUAL_VALUE; // //ACTUAL VALUE represents real physical value returned by READ SINGLE REGISTER REQUEST
```
Comment (kodsurf) [inline src/mod/l4_thruster.h:23]: Idea was to use is as "OBC registers" that would store latest parsed values from read request

File: src/mod/thr/thr.c [hunk 6]
```diff
@@ -217,11 +217,10 @@ void thrProcessRxByte(uint8_t rxByte) {
 	}
 	else {
 		l4_thr_counter =0;
-		//print whole receive buffer
-		//SysEvent(MODULE_ID_CLIMBAPP, EVENT_INFO, EID_APP_STRING, thr_receiveBuffer, strlen(thr_receiveBuffer));
-
-		//thr_debugPrintBuffer( (uint8_t*)thr_receiveBuffer,strlen(thr_receiveBuffer));
-		ParseReadRequest((uint8_t*)thr_receiveBuffer,l4_thr_ExpectedReceiveBuffer);
+		// This was not yet tested on OBC ! But works on standaone .c
+		ParseReadRequest((uint8_t*)thr_receiveBuffer,l4_thr_ExpectedReceiveBuffer); //
```
Comment (kodsurf) [inline src/mod/thr/thr.c:184]: I have not yet tested that it works correctly on OBC !!!!

**Review by RobertK66 — DISMISSED** [review]
mod/crc/obc_checksums
---------------------------

I moved (parts of) this code some time ago to ado lib:
ado-chip-175x-6x\ado_chip_175x_6x\src\ado_crc.c

Pls. use the code from there and integrate changes you need for your module there.


.. review t.b.c. ...
(not sure how to assign comments to code files/positions yet ;-)....

File: src/mod/crc/obc_checksums.c [hunk 7]
```diff
@@ -0,0 +1,122 @@
+#include <stdio.h>
```
Comment (RobertK66) [inline src/mod/crc/obc_checksums.c:1]: Ah,you copied this from hardwaretest repos...

My intention of crc was to get common usable code into:
ado-chip-175x-6x\ado_chip_175x_6x\src\ado_crc.c (*.h)

Not sure yet what you added used here but will comment it there.

File: src/ClimbObc.c [hunk 8]
```diff
@@ -73,6 +75,12 @@ static gps_initdata_t GpsInit = {
 		PTR_FROM_IDX(PINIDX_STACIE_C_IO1_P)
 };
 
+static thr_initdata_t ThrInit = {
+		LPC_UART1, /// WILL USE UART1 since it is on Y+ side, which according to doccumentation given to me should be used for RS485 thruster LPC_UART1
+		PTR_FROM_IDX(PINIDX_GPIO4_CP),
```
Comment (RobertK66) [inline src/ClimbObc.c:80]: Do you use this additional Pins for thruster controll. If so pls. comment here which pins with which functions. (init structure was copied from GPS and there where pps and enable pin...)

File: src/ClimbObc.c [hunk 9]
```diff
@@ -94,7 +102,10 @@ static const MODULE_DEF_T Modules[] = {
 		MOD_INIT( sen_init, sen_main, NULL),
 		MOD_INIT( memInit, memMain, &MemoryInit),
 		MOD_INIT( gpsInit, gpsMain, &GpsInit),
-		MOD_INIT( app_init, app_main, NULL)
+		MOD_INIT( app_init, app_main, NULL),
+		MOD_INIT( thrInit, thrMain, &ThrInit),
+		MOD_INIT( app_init, app_main, NULL),
```
Comment (RobertK66) [inline src/ClimbObc.c:107]: seems to be duplicated line (from merge !?) -> pls. delete

File: src/mod/l2_debug_com.h [hunk 10]
```diff
@@ -15,7 +15,7 @@
 #include "ado_modules.h"
 
 // RX definitions
-#define DEB_L2_MAX_CHARPERLINE		100
+#define DEB_L2_MAX_CHARPERLINE		10
```
Comment (RobertK66) [inline src/mod/l2_debug_com.h:18]: Was this changed intentional!?  I remember this is the input buffer for a single command line and there should be more room than 10 char, especially when we allow 10 args (with the next define!) ??

File: src/mod/l4_thruster.c [hunk 11]
```diff
@@ -0,0 +1,539 @@
+/*
```
Comment (RobertK66) [inline src/mod/l4_thruster.c:1]: I gave you general review on the thruster globally with mail. So for having a faster pull request to develop I mainly skip this part(s) for now. I assume it works and does not block mainloop and other functionality. 
We can make cleanup/rerfactorings for memory or other optimizations later.

File: src/mod/l4_thruster.c [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+		request[1] = DEVICE;
+		request[2] = MSGTYPE[2]; // READ -3
+		request[3] = 0x00; // checksum
+		request[4] = 0x02; // LENGTH of payload REGISTER and length
+		request[5] = 0x00; // Hardcoded because length of payload is defined with two bytes of uint16
+		request[6]= access_register; // Access register at address requested by used
+		request[7]= length_of_register;
+
+		uint8_t len = sizeof(request);
+		request[3] = CRC8(request,len);
```
Comment (RobertK66) [inline src/mod/l4_thruster.c:528]: not sure the CRC8() coded in ado_crc is exactly th8is polinom (you copied over from hardwaretest repos). If yes pls. use ado function. If not we should make the CRC8 implementation in ado fitter .....

File: src/mod/thr/thr.c [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+	// SET TO RECEIVE UPON INITIALIZATION
+	//Chip_GPIO_SetPinOutLow(LPC_GPIO, 2, 5); //  This is PINIDX_RS485_TX_RX RECEIVE
+	//Chip_GPIO_SetPinOutHigh(LPC_GPIO, 2, 5); //  This is PINIDX_RS485_TX_RX TRANSMIT
+
+	// Init UART
+	InitUart(thrInitData->pUart, 115200, thrUartIRQ);
+	thrFirstByteAfterReset = true;
+
+	// Enable dirrection controll
+	LPC_UART1->RS485CTRL |= UART_RS485CTRL_DCTRL_EN; // Enable Auto Direction Control
```
Comment (RobertK66) [inline src/mod/thr/thr.c:87]: here and next lines  you should use the configured uart and not a hardcoded one !

File: src/mod/thr/thr.c [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+	}
+
+}
+
+
+
+
+
+void thr_debugPrintBuffer(uint8_t *buffer,int bufferlen){
```
Comment (RobertK66) [inline src/mod/thr/thr.c:196]: again never use this and pls. remove

File: src/mod/thr/thr.h [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+//
+#ifndef MOD_TIM_CLIMB_THR_H_
+#define MOD_TIM_CLIMB_THR_H_
+
+#include <chip.h>
+
+// init data needed. Choose UARt and 2 GPIO Pins to be used
+typedef struct {
+	LPC_USART_T 	*pUart;			// default is 9600baud
+	const PINMUX_GRP_T2   *pEnablePin;	// Output LPC -> GPS enable internal Voltage regulator -> we disable because we use SP Power
```
Comment (RobertK66) [inline src/mod/thr/thr.h:17]: remove this old io definiotions (copied over from GPS)

**Review by RobertK66 — CHANGES_REQUESTED** [review]
For me it looks ok to get into devlop. Only 'major' thing on that level is the usage of CRC8() via ado_crc implementation. I only understood CRC math when I was confronted with the 'CRC16 families' I had to use with comms.....
For CRC8 I am not sure if current implementation gives enough flexibility to be used as is......
Pls. remove the OBC_Chechsums.c/h and try to use ado_crc. If you do need your own version(s) then make a temp routine somewhere and we together move merge it into ado_crc afterwards.

**Comment by RobertK66** [comment]
I just made commit https://github.com/RobertK66/obc_1769_core/pull/34/commits/c1f7512bea464726c193a5ec93afe598cd3d0a43 in order to get the project compiled again ;-)

File: src/mod/crc/obc_checksums.c (hunk 7 above)
Comment (kodsurf) [reply src/mod/crc/obc_checksums.c:1]: Yes, I copied from old repo ?

I can look through ado-chip and change to it if needed

File: src/mod/l2_debug_com.h (hunk 10 above)
Comment (kodsurf) [reply src/mod/l2_debug_com.h:18]: This was not mine change. I believe it was from develop branch.

Comment (RobertK66) [reply src/mod/l2_debug_com.h:18]: ok never mind where it comes from. Change back to 100 pls.

File: src/mod/crc/obc_checksums.c (hunk 7 above)
Comment (kodsurf) [reply src/mod/crc/obc_checksums.c]: Checksum calculated with 
image

image

is **wrong** . Therefore thruster did not worked.

My solution with commit :

https://github.com/RobertK66/obc_1769_core/commit/d5abfb96bfd8000728830127751fe0fd10528da7

Would be to copy paste correct checksum function (that works with thruster) into ado_crc.c


I dont have access rights for ado lib. Thus I will send updated files to your email.  Please  commit.

File: src/ClimbObc.c (hunk 9 above)
Comment (kodsurf) [reply src/ClimbObc.c]: https://github.com/RobertK66/obc_1769_core/pull/34/commits/0e793dce1459ebc0f6b293d8c4a6936331faa6b2

Fixed

File: src/mod/l2_debug_com.h (hunk 10 above)
Comment (kodsurf) [reply src/mod/l2_debug_com.h]: https://github.com/RobertK66/obc_1769_core/pull/34/commits/841628fe94e1916b2ffdb5d92fc97200093c1b6a

Fixed

File: src/mod/thr/thr.c (hunk 13 above)
Comment (kodsurf) [reply src/mod/thr/thr.c]: https://github.com/RobertK66/obc_1769_core/pull/34/commits/34ed796648ff98ced677495b560fa473a2707849

Fixed

File: src/mod/thr/thr.h (hunk 15 above)
Comment (kodsurf) [reply src/mod/thr/thr.h]: Fixed

https://github.com/RobertK66/obc_1769_core/pull/34/commits/536712ccc92806a2e5c117fd2d5c82f9ab6059e1

File: src/mod/crc/obc_checksums.c (hunk 7 above)
Comment (RobertK66) [reply src/mod/crc/obc_checksums.c]: I would sugest you move the crc_thruster() into thruster.c in order to make the code checked in compilable (merging to develop will generate a build and tries to make the image!).
I will move it to ado from there on later!

**Review by RobertK66 — CHANGES_REQUESTED** [review]
looks good now. 2 (minor) points left.
- 1st:  The current code does not compile (because of your request to move crc8_thruster to ado!). I propose to include the code locally in thruster for this pull request now and then I will move it to ado in e second step afterwards.
- 2nd:  Pls. also add the 3 new commands in the table in doc/m01_debug_if.md file. -> this will get 'compiled' into the PDF OBC manual if we release a version.

**Comment by RobertK66** [comment]
@kodsurf : are you available at skype. We could finalise this there with shared screen if you wish....

**Comment by kodsurf** [comment]
With latest commit I made thruster parser work as expected. 


Parser is tested.  

I now approve merge to dev

**Review by RobertK66 — APPROVED** [review]
Fine I think that's it for now

**Pull request merged**